storage/client/credentials.cc

```diff
@@ -0,0 +1,52 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/client/credentials.h"
+#include "storage/client/internal/service_account_credentials.h"
+#include <fstream>
+#include <iostream>
+#include <iterator>
+
+namespace {
+#ifdef _WIN32
+char const CREDENTIALS_ENV_VAR[] = "APPDATA";
+#else
+char const CREDENTIALS_ENV_VAR[] = "HOME";
+#endif
+
+std::string const& GoogleCredentialsSuffix() {
+#ifdef _WIN32
+  static std::string const suffix =
+      "/gcloud/application_default_credentials.json";
+#else
+  static std::string const suffix =
+      "/.config/gcloud/application_default_credentials.json";
+#endif
+  return suffix;
+}
+}  // anonymous namespace
+
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+std::shared_ptr<Credentials> GoogleDefaultCredentials() {
+  // There are probably more efficient ways to do this, but meh, the strings
```

Comment:
Similar to the order checked in other client libraries (looking at "GetApplicationDefaultAsync()" in [1]), can we check for the GOOGLE_APPLICATION_CREDENTIALS env var first, and then fall back to the gcloud ADC file?

Also, `root` ends up being `NULL` if "HOME" is not a defined env var for some weird reason (e.g. see [2]).

[1] https://developers.google.com/api-client-library/dotnet/reference/1.9.3/classGoogle_1_1Apis_1_1Auth_1_1OAuth2_1_1GoogleCredential
[2] https://github.com/angular/angular-cli/issues/6806

Reply:
On `GOOGLE_APPLICATION_CREDENTIALS`:  yes, but that is what you are going to design and implement, I am just implementing the easy part to get going :grin:  Created #522, but that is one of many other things to do, for example, also need to use GCE metadata if available.

On `HOME` not being defined: great point, thanks created #523, we can fix in this PR on a future one, up to you.

Reply:
Ah, gotcha :D If this is only intended as a placeholder to provide actual credentials functionality (and only operates against gcloud's ADC file), could you leave a comment/warning to that effect at the top of the class?